File: next/Kernel/Outcome.fs
namespace Wanxiangshu.Next.Kernel

open Wanxiangshu.Next.Kernel.Identity

/// Typed completion payload for a successful agent run.
/// Carries the completion payload for a successful agent run.
/// No transport parts are kept; business programs inspect the typed turn directly.
type AgentRunResult =
    { SessionId: SessionId
      RootUserMessageId: MessageId
      AssistantMessageId: MessageId
      Role: string // AgentRole serialized as string (avoid Kernel→Session dep)
      Directory: string
      /// Session-wide A: cumulative formal text + reasoning/thinking across the
      /// whole Session, not the last turn alone. Excludes tool raw streams.
      FinalText: string
      /// Current turn's formal assistant text only (no reasoning/thinking).
      /// B record for blogger companions is built from this field.
      FormalText: string }

    /// Hard invariant: completed runs must have non-empty session-wide A text.
    member this.IsValid = not (System.String.IsNullOrWhiteSpace this.FinalText)

type AgentRunFailure =
    { SessionId: SessionId; Reason: string }


module Outcome =

    type SendOutcome =
        | Delivered of MessageId
        | Retryable of reason: string
        | AcceptanceUnknown of reason: string * messageId: MessageId option
        | Fatal of reason: string

    type SessionOutcome =
        | CompletedSession of message: string
        | CancelledSession
        | TerminatedSession of reason: string

    type SessionError =
        | NoProgress of reason: string
        | SessionCancelled
        | FallbackExhausted
        | ReviewExhausted
        | PromptUncertain
        | ProjectionBroken of reason: string
        | InboxFull
        | Protocol of reason: string

    type JournalFailure =
        | WriteFailed of reason: string
        | FlushFailed of reason: string

    type CommitResult<'e> =
        | Committed of 'e
        | CommitUnknown of EventId * JournalFailure

File: .wanxiangshu/runtimes/Kernel/Outcome.js

import { Union } from "../fable_modules/fable-library-js.5.6.0/Types.js";
import { EventId_$reflection, MessageId_$reflection } from "./Identity.js";
import { union_type, option_type, string_type } from "../fable_modules/fable-library-js.5.6.0/Reflection.js";

export class SendOutcome extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Delivered", "Retryable", "AcceptanceUnknown", "Fatal"];
    }
}

export function SendOutcome_$reflection() {
    return union_type("Wanxiangshu.Next.Kernel.Outcome.SendOutcome", [], SendOutcome, () => [[["Item", MessageId_$reflection()]], [["reason", string_type]], [["reason", string_type], ["messageId", option_type(MessageId_$reflection())]], [["reason", string_type]]]);
}

export class SessionOutcome extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["CompletedSession", "CancelledSession", "TerminatedSession"];
    }
}

export function SessionOutcome_$reflection() {
    return union_type("Wanxiangshu.Next.Kernel.Outcome.SessionOutcome", [], SessionOutcome, () => [[["message", string_type]], [], [["reason", string_type]]]);
}

export class SessionError extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["NoProgress", "SessionCancelled", "FallbackExhausted", "ReviewExhausted", "PromptUncertain", "ProjectionBroken", "InboxFull", "Protocol"];
    }
}

export function SessionError_$reflection() {
    return union_type("Wanxiangshu.Next.Kernel.Outcome.SessionError", [], SessionError, () => [[["reason", string_type]], [], [], [], [], [["reason", string_type]], [], [["reason", string_type]]]);
}

export class JournalFailure extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["WriteFailed", "FlushFailed"];
    }
}

export function JournalFailure_$reflection() {
    return union_type("Wanxiangshu.Next.Kernel.Outcome.JournalFailure", [], JournalFailure, () => [[["reason", string_type]], [["reason", string_type]]]);
}

export class CommitResult$1 extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Committed", "CommitUnknown"];
    }
}

export function CommitResult$1_$reflection(gen0) {
    return union_type("Wanxiangshu.Next.Kernel.Outcome.CommitResult`1", [gen0], CommitResult$1, () => [[["Item", gen0]], [["Item1", EventId_$reflection()], ["Item2", JournalFailure_$reflection()]]]);
}

